base/base/sort.h

```diff
@@ -2,6 +2,90 @@
 
 #include <pdqsort.h>
 
+#ifdef NDEBUG
+
+#include <pcg_random.hpp>
+#include <base/getThreadId.h>
+
+/** Same as libcxx std::__debug_less. Just without dependency on private part of standard library.
+  * Check that Comparator induce strict weak ordering.
+  */
+template <typename Comparator>
+class DebugLessComparator
+{
+public:
+    constexpr DebugLessComparator(Comparator & cmp_)
+        : cmp(cmp_)
+    {}
+
+    template <typename LhsType, typename RhsType>
+    constexpr bool operator()(const LhsType & lhs, const RhsType & rhs)
+    {
+        bool lhs_less_than_rhs = cmp(lhs, rhs);
+        if (lhs_less_than_rhs)
+            assert(!cmp(rhs, lhs));
+
+        return lhs_less_than_rhs;
+    }
+
+    template <typename LhsType, typename RhsType>
+    constexpr bool operator()(LhsType & lhs, RhsType & rhs)
+    {
+        bool lhs_less_than_rhs = cmp(lhs, rhs);
+        if (lhs_less_than_rhs)
+            assert(!cmp(rhs, lhs));
+
+        return lhs_less_than_rhs;
+    }
+
+private:
+    Comparator & cmp;
+};
+
+template <typename Comparator>
+using ComparatorWrapper = DebugLessComparator<Comparator>;
+
+template <typename RandomIt, typename Compare>
+void shuffleItemsInEqualRanges(RandomIt first, RandomIt last, Compare compare)
```

Comment:
Why do you need such complex logic?

Shuffle before sort makes the probabilities exactly the same

For partial sorts shuffle before sort+shuffle of the remaining part after, for nth_element shuffle before the call + 2 shuffles on both sides